// @flowmap.gl/layers/dist/AnimatedFlowLinesLayer/AnimatedFlowLinesLayer.js
/*
 * Copyright (c) Flowmap.gl contributors
 * Copyright (c) 2018-2020 Teralytics
 * SPDX-License-Identifier: Apache-2.0
 */
import { Layer, picking, project32 } from '@deck.gl/core';
import GL from '@luma.gl/constants';
import { Geometry, Model } from '@luma.gl/core';
import FragmentShader from './AnimatedFlowLinesLayerFragment.glsl';
import VertexShader from './AnimatedFlowLinesLayerVertex.glsl';
const DEFAULT_COLOR = [0, 132, 193, 255];
const loopLength = 1800;
const animationSpeed = 20;
const loopTime = loopLength / animationSpeed;
class AnimatedFlowLinesLayer extends Layer {
    constructor(props) {
        super(props);
    }
    getShaders() {
        return super.getShaders({
            vs: VertexShader,
            fs: FragmentShader,
            modules: [project32, picking],
            shaderCache: this.context.shaderCache,
        });
    }
    initializeState() {
        const attributeManager = this.getAttributeManager();
        /* eslint-disable max-len */
        attributeManager.addInstanced({
            instanceSourcePositions: {
                size: 3,
                type: GL.DOUBLE,
                transition: true,
                accessor: 'getSourcePosition',
            },
            instanceTargetPositions: {
                size: 3,
                type: GL.DOUBLE,
                transition: true,
                accessor: 'getTargetPosition',
            },
            instanceColors: {
                size: 4,
                type: GL.UNSIGNED_BYTE,
                transition: true,
                accessor: 'getColor',
                defaultValue: [0, 0, 0, 255],
            },
            instanceWidths: {
                size: 1,
                transition: true,
                accessor: 'getThickness',
                defaultValue: 1,
            },
            instanceStaggering: {
                accessor: 'getStaggering',
                size: 1,
                transition: false,
            },
            instancePickable: {
                accessor: 'getPickable',
                size: 1,
                transition: false,
            },
        });
        /* eslint-enable max-len */
    }
    getNeedsRedraw() {
        return true;
    }
    updateState({ props, oldProps, changeFlags }) {
        super.updateState({ props, oldProps, changeFlags });
        if (changeFlags.extensionsChanged) {
            const { gl } = this.context;
            this.state.model?.delete();
            this.state.model = this._getModel(gl);
            this.getAttributeManager().invalidateAll();
        }
    }
    draw({ uniforms }) {
        const { thicknessUnit, animationTailLength } = this.props;
        const timestamp = Date.now() / 1000;
        const animationTime = ((timestamp % loopTime) / loopTime) * loopLength;
        this.state.model
            .setUniforms({
            ...uniforms,
            thicknessUnit: thicknessUnit * 4,
            animationTailLength,
            currentTime: animationTime,
        })
            .draw();
    }
    _getModel(gl) {
        /*
         *  (0, -1)-------------_(1, -1)
         *       |          _,-"  |
         *       o      _,-"      o
         *       |  _,-"          |
         *   (0, 1)"-------------(1, 1)
         */
        const positions = [0, -1, 0, 0, 1, 0, 1, -1, 0, 1, 1, 0];
        return new Model(gl, Object.assign({}, this.getShaders(), {
            id: this.props.id,
            geometry: new Geometry({
                drawMode: GL.TRIANGLE_STRIP,
                attributes: {
                    positions: new Float32Array(positions),
                },
            }),
            isInstanced: true,
        }));
    }
}
AnimatedFlowLinesLayer.defaultProps = {
    currentTime: 0,
    animationTailLength: 0.7,
    getSourcePosition: { type: 'accessor', value: (d) => [0, 0] },
    getTargetPosition: { type: 'accessor', value: (d) => [0, 0] },
    getPickable: { type: 'accessor', value: (d) => 1.0 },
    getStaggering: {
        type: 'accessor',
        value: (d, { index }) => Math.random(),
    },
    getColor: { type: 'accessor', value: DEFAULT_COLOR },
    getThickness: { type: 'accessor', value: 1 },
    thicknessUnit: 15 * 2,
    parameters: {
        depthTest: false,
    },
};
export default AnimatedFlowLinesLayer;
