// @flowmap.gl/layers/dist/FlowLinesLayer/FlowLinesLayer.js
/*
 * Copyright (c) Flowmap.gl contributors
 * Copyright (c) 2018-2020 Teralytics
 * SPDX-License-Identifier: Apache-2.0
 */
import { Layer, picking, project32 } from '@deck.gl/core';
import GL from '@luma.gl/constants';
import { Geometry, Model } from '@luma.gl/core';
import FragmentShader from './FlowLinesLayerFragment.glsl';
import VertexShader from './FlowLinesLayerVertex.glsl';
const DEFAULT_COLOR = [0, 132, 193, 255];
const INNER_SIDE_OUTLINE_THICKNESS = 1;
// source_target_mix, perpendicular_offset_in_thickness_units, direction_of_travel_offset_in_thickness_units
// prettier-ignore
const POSITIONS = [
    1, 0, 0,
    1, 2, -3,
    1, 1, -3,
    1, 0, 0,
    1, 1, -3,
    0, 1, 0,
    1, 0, 0,
    0, 1, 0,
    0, 0, 0, // 4
];
/**
                                    1
                                    ··
                                    · ··
                                    ·    ··
     3                            2 ·      ··
      ·······························        ··
      · ·······                       ····      ··
      ·       ········                   ·····     ··
      ·               ···············        ·····   ··
      ·                             ········      ········
      ·                                     ················
    4 ························································  0

 */
function getOutlinePixelOffsets(tout, tin) {
    // perpendicular_offset_in_pixels, direction_of_travel_offset_in_pixels, fill_outline_color_mix
    // prettier-ignore
    return ([
        -tin, 2 * tout, 1,
        2 * tout, -tout, 1,
        tout, -tout, 1,
        -tin, 2 * tout, 1,
        tout, -tout, 1,
        tout, -tout, 1,
        -tin, 2 * tout, 1,
        tout, -tout, 1,
        -tin, -tout, 1, // 4
    ]);
}
// prettier-ignore
const ZEROES = [
    0, 0, 0,
    0, 0, 0,
    0, 0, 0,
    0, 0, 0,
    0, 0, 0,
    0, 0, 0,
    0, 0, 0,
    0, 0, 0,
    0, 0, 0,
];
class FlowLinesLayer extends Layer {
    // props!: Props;
    constructor(props) {
        super(props);
    }
    getShaders() {
        return super.getShaders({
            vs: VertexShader,
            fs: FragmentShader,
            modules: [project32, picking],
            shaderCache: this.context.shaderCache,
        });
    }
    initializeState() {
        const { attributeManager } = this.state;
        attributeManager.addInstanced({
            instanceSourcePositions: {
                accessor: 'getSourcePosition',
                size: 3,
                transition: false,
                type: GL.DOUBLE,
            },
            instanceTargetPositions: {
                accessor: 'getTargetPosition',
                size: 3,
                transition: false,
                type: GL.DOUBLE,
            },
            instanceThickness: {
                accessor: 'getThickness',
                size: 1,
                transition: false,
            },
            instanceEndpointOffsets: {
                accessor: 'getEndpointOffsets',
                size: 2,
                transition: false,
            },
            instanceColors: {
                accessor: 'getColor',
                size: 4,
                type: GL.UNSIGNED_BYTE,
                transition: false,
            },
            instancePickable: {
                accessor: 'getPickable',
                size: 1,
                transition: false,
            },
        });
    }
    updateState({ props, oldProps, changeFlags }) {
        super.updateState({ props, oldProps, changeFlags });
        if (changeFlags.extensionsChanged) {
            const { gl } = this.context;
            if (this.state.model) {
                this.state.model.delete();
            }
            this.setState({ model: this._getModel(gl) });
            this.getAttributeManager().invalidateAll();
        }
    }
    draw({ uniforms }) {
        const { gl } = this.context;
        const { outlineColor, thicknessUnit } = this.props;
        gl.lineWidth(1);
        this.state.model
            .setUniforms({
            ...uniforms,
            outlineColor: outlineColor.map((x) => x / 255),
            // outlineColor: [1, 0, 0, 1],
            thicknessUnit: thicknessUnit * 2.0,
            gap: 0.5,
        })
            .draw();
    }
    _getModel(gl) {
        let positions = [];
        let pixelOffsets = [];
        const { drawOutline, outlineThickness } = this.props;
        if (drawOutline) {
            // source_target_mix, perpendicular_offset_in_thickness_units, direction_of_travel_offset_in_thickness_units
            positions = positions.concat(POSITIONS);
            const tout = outlineThickness;
            const tin = INNER_SIDE_OUTLINE_THICKNESS; // the outline shouldn't cover the opposite arrow
            pixelOffsets = pixelOffsets.concat(getOutlinePixelOffsets(tout, tin));
        }
        positions = positions.concat(POSITIONS);
        pixelOffsets = pixelOffsets.concat(ZEROES);
        return new Model(gl, {
            id: this.props.id,
            ...this.getShaders(),
            geometry: new Geometry({
                drawMode: GL.TRIANGLES,
                attributes: {
                    positions: new Float32Array(positions),
                    normals: new Float32Array(pixelOffsets),
                },
            }),
            isInstanced: true,
            shaderCache: this.context.shaderCache,
        });
    }
}
FlowLinesLayer.layerName = 'FlowLinesLayer';
FlowLinesLayer.defaultProps = {
    getSourcePosition: { type: 'accessor', value: (d) => [0, 0] },
    getTargetPosition: { type: 'accessor', value: (d) => [0, 0] },
    getColor: { type: 'accessor', value: DEFAULT_COLOR },
    getThickness: { type: 'accessor', value: (d) => d.count },
    getPickable: { type: 'accessor', value: (d) => 1.0 },
    drawOutline: true,
    thicknessUnit: 12,
    outlineThickness: 1,
    outlineColor: [255, 255, 255, 255],
    parameters: {
        depthTest: false,
    },
};
export default FlowLinesLayer;
